lib/Sema/CSSimplify.cpp

```diff
@@ -5969,6 +5969,60 @@ ConstraintSystem::simplifyApplicableFnConstraint(
     return simplified;
   }
 
+  // Handle applications of types with `callAsFunction` methods.
+  if (desugar2->mayHaveMembers()) {
```

Comment:
In addition to that it would be great to add a new path element for this kind of situation, something like `implicit call` so it could easy easily identified and diagnosed.

Reply:
Thanks for your suggestions @xedin! I explored your suggestions and got stuck (`mutating func callAsFunction` doesn't work with member constraint) - posted info in [this comment](https://github.com/dan-zheng/swift/commit/394c589198288f41c64012e182bc85faa530354c#commitcomment-34646601). Could you please take a look? 🙂

Reply:
Happy to help! I have replied on that comment, it should be pretty easy to fix.

Reply:
`func callAsFunction` constraint revamp done in https://github.com/apple/swift/pull/24299/commits/8dcd1c68b2acbf6fffc664b853b168a851584ea5 - please read the commit description for more details.

@xedin: could you please re-review?